python/lsst/daf/butler/cli/log.py

```diff
@@ -0,0 +1,160 @@
+    normalLogFmt = "%c %p: %m%n"
+
+    @staticmethod
+    def getPyLogLevel(level):
+        """Get the numeric value for the given log level name.
+
+        Parameters
+        ----------
+        level : `str`
+            One of the python `logging` log level names.
+
+        Returns
+        -------
+        numericValue : `int`
+            The python `logging` numeric value for the log level.
+        """
+        return getattr(logging, level, None)
+
+    @staticmethod
+    def getLsstLogLevel(level):
+        """Get the numeric value for the given log level name.
+
+        If `lsst.log` is not setup this function will return `None` regardless
+        of input. `daf_butler` does not directly depend on `lsst.log` and so it
+        will not be setup when `daf_butler` is setup. Packages that depend on
+        `daf_butler` and use `lsst.log` may setup `lsst.log`.
+
+        Will adapt the python name to an `lsst.log` name:
+        - CRITICAL to FATAL
+        - WARNING to WARN
+
+        Parameters
+        ----------
+        level : `str`
+            One of the python `logging` log level names.
+
+        Returns
+        -------
+        numericValue : `int` or `None`
+            The `lsst.log` numeric value.
+        """
+        if lsstLog is None:
+            return None
+        if level == "CRITICAL":
+            level = "FATAL"
+        elif level == "WARNING":
+            level = "WARN"
+        return getattr(lsstLog.Log, level, None)
+
+    @classmethod
+    def initLog(cls, longlog):
+        """Initialize logging. This should only be called once per program
+        execution. After the first call this will log a warning and return.
+
+        Parameters
+        ----------
+        longlog : `bool`
+            If True, make log messages appear in long format, by default False.
+        """
+        if cls._initialized:
+            # Unit tests that execute more than one command do end up
+            # calling this fucntion multiple times in one program execution,
+            # so do warn but don't log an error or fail, just make the
+            # re-initalization a no-op.
+            log = logging.getLogger(__name__.partition(".")[2])
+            log.warning("Log is already initialized, returning without re-initializing.")
+            return
+        cls._initialized = True
+
+        lgr = logging.getLogger()
+        if lsstLog is not None:
+            # global logging config
+
+            lsstLog.configure_prop(_LOG_PROP.format(cls.longLogFmt if longlog else cls.normalLogFmt))
+
+            # Forward all Python logging to lsstLog
+            lgr.addHandler(lsstLog.LogHandler())
+        lgr.setLevel(logging.INFO)
+
+        # also capture warnings and send them to logging
+        logging.captureWarnings(True)
+
+    @classmethod
+    def uninitLog(cls, components):
+        if lsstLog is not None:
+            lgr = logging.getLogger()
+            lgr.removeHandler(lsstLog.LogHandler())
```

Comment:
Hmm 🤔 I'm not sure it is going to work. Do you have unit test for this?

Reply:
@andy-slac can you say more about what part won't work?

Reply:
What I think is that `removeHandler` removes existing handler from the the handler list. `lsstLog.LogHandler()` make a new handler that was never added to that list, so removing it does not even make sense (it is probably no-op). What you want is to remove handler instance that you created earlier and added with `addHandler` , this means you need to remember that instance somewhere. As for testing, I think you can check the size of the `lgr.handlers` to see if removing changes its size (though `handlers` is an undocumented attribute)

Reply:
You can't test in daf_butler itself because daf_butler doesn't depend on lsst.log.

Reply:
You can if you (accidentally) setup `log` or some other package that depends on `log`.

Reply:
That is true but you can't rely on it from a Jenkins point of view. ctrl_mpexec can test it though.

Reply:
I have tests in obs_base for features in daf_butler that depend on lsst.log. Let me know if you think that's not a good place.
I don't think I can depend on the actual number of handlers, tests may run before this test may affect it. I'm not even sure I can depend on a delta. I might have to figure out how to tell if a specific handler is in the list, or punt on being able to know.
The affect that persistent logging configuration has on unrelated unit tests that run in the same process is really giving me hell. I might have questions in Slack today.

Reply:
obs_base is good.   Logging does affect global state which is why logger configurations should only be triggered by command line tooling and not by module code.